awx/main/tests/functional/test_inventory_source_injectors.py

```diff
@@ -0,0 +1,294 @@
+
+    def substitute_run(args, cwd, call_env, stdout_handle, **_kw):
+        """This method will replace run_pexpect
+        instead of running, it will read the private data directory contents
+        It will make assertions that the contents are correct
+        If MAKE_INVENTORY_REFERENCE_FILES is set, it will produce reference files
+        """
+        private_data_dir = call_env.pop('AWX_PRIVATE_DATA_DIR')
+        set_files = bool(os.getenv("MAKE_INVENTORY_REFERENCE_FILES", 'false').lower()[0] not in ['f', '0'])
+        env, content = read_content(private_data_dir, call_env, inventory_update)
+        base_dir = os.path.join(DATA, script_or_plugin)
+        if not os.path.exists(base_dir):
+            os.mkdir(base_dir)
+        source_dir = os.path.join(base_dir, this_kind)  # this_kind is a global
+        if set_files:
+            create_reference_data(source_dir, env, content)
+            pytest.skip('You set MAKE_INVENTORY_REFERENCE_FILES, so this created files, unset to run actual test.')
+        else:
+            if not os.path.exists(source_dir):
+                raise FileNotFoundError(
+                    'Maybe you never made reference files? '
+                    'MAKE_INVENTORY_REFERENCE_FILES=true py.test ...\noriginal: {}')
+            files_dir = os.path.join(source_dir, 'files')
+            try:
+                expected_file_list = os.listdir(files_dir)
+            except FileNotFoundError:
+                expected_file_list = []
+            assert set(expected_file_list) == set(content.keys()), (
+                'Inventory update runtime environment does not have expected files'
+            )
+            for f_name in expected_file_list:
+                with open(os.path.join(files_dir, f_name), 'r') as f:
+                    ref_content = f.read()
+                    assert content[f_name] == ref_content
+            try:
+                with open(os.path.join(source_dir, 'env.json'), 'r') as f:
+                    ref_env_text = f.read()
+                    ref_env = json.loads(ref_env_text)
+            except FileNotFoundError:
+                ref_env = {}
+            assert env == ref_env
+        return ('successful', 0)
+
+    # Mock this so that it will not send events to the callback receiver
+    # because doing so in pytest land creates large explosions
+    with mock.patch('awx.main.queue.CallbackQueueDispatcher.dispatch', lambda self, obj: None):
+        # Force the update to use the script injector
+        with mock.patch('awx.main.models.inventory.PluginFileInjector.should_use_plugin', return_value=use_plugin):
+            # Also do not send websocket status updates
+            with mock.patch.object(UnifiedJob, 'websocket_emit_status', mock.Mock()):
+                # The point of this test is that we replace run_pexpect with assertions
+                with mock.patch('awx.main.expect.run.run_pexpect', substitute_run):
```

Comment:
You don't have to quadruple-nest these context managers:

https://docs.python.org/2/library/contextlib.html#contextlib.nested

Reply:
ahh did not know that

Reply:
There's a lot of unlearning I have to do from making things python2 compatible. But for these, I also don't like formatting the indents before the `:`. I just can't make it look good no matter what I do.

Reply:
Also can do something like move it up to the decorator:
https://github.com/ansible/ansible-runner/blob/33c7c7689388c9c3c0f68a81c298dcc08fbc1dcf/test/unit/test_runner_config.py#L410-L413